src/psfmachine/perturbation.py

```diff
@@ -0,0 +1,576 @@
+
+    @property
+    def nvec(self):
+        return self.vectors.shape[1]
+
+    @property
+    def ntime(self):
+        return self.time.shape[0]
+
+    def bin_func(self, var, **kwargs):
+        """
+        Bins down an input variable to the same time resolution as `self`
+
+        Parameters
+        ----------
+        var: npt.ArrayLike
+            Array of values with at least 1 dimension. The first dimension must be
+            the same shape as `self.time`
+        """
+        if self.bin_method.lower() == "downsample":
+            func = self._get_downsample_func()
+        elif self.bin_method.lower() == "bin":
+            func = self._get_bindown_func()
+        else:
+            raise NotImplementedError
+        return func(var, **kwargs)
+
+    def _get_downsample_func(self):
+        """Builds a function to lower the resolution of the data through downsampling"""
+        points = []
+        b = np.hstack([0, self.breaks, len(self.time) - 1])
+        for b1, b2 in zip(b[:-1], b[1:]):
+            p = np.arange(b1, b2, self.resolution)
+            if p[-1] != b2:
+                p = np.hstack([p, b2])
+            points.append(p)
+        points = np.unique(np.hstack(points))
+        self.nbins = len(points)
+
+        def func(x, quad=False):
+            """
+            Bins down an input variable to the same time resolution as `self`
+            """
+            if x.shape[0] == len(self.time):
+                return x[points]
+            else:
+                raise ValueError("Wrong size to bin")
+
+        return func
+
+    def _get_bindown_func(self):
+        """Builds a function to lower the resolution of the data through binning"""
+        b = np.hstack([0, self.breaks, len(self.time) - 1])
+        points = np.hstack(
+            [np.arange(b1, b2, self.resolution) for b1, b2 in zip(b[:-1], b[1:])]
+        )
+        points = points[~np.in1d(points, np.hstack([0, len(self.time) - 1]))]
+        points = np.unique(np.hstack([points, self.breaks]))
+        self.nbins = len(points) + 1
+
+        def func(x, quad=False):
+            """
+            Bins down an input variable to the same time resolution as `self`
+            """
+            if x.shape[0] == len(self.time):
+                if not quad:
+                    return np.asarray(
+                        [i.mean(axis=0) for i in np.array_split(x, points)]
+                    )
+                else:
+                    return (
+                        np.asarray(
+                            [
+                                np.sum(i ** 2, axis=0) / (len(i) ** 2)
+                                for i in np.array_split(x, points)
+                            ]
+                        )
+                        ** 0.5
+                    )
+            else:
+                raise ValueError("Wrong size to bin")
+
+        return func
+
+    def pca(self, y, ncomponents=5, smooth_time_scale=0):
```

Comment:
to be consistent with the rest

```suggestion
    def pca(self, y, ncomponents=3, smooth_time_scale=0):
```